---
title: "Example Uses of snps2assays"
author: "Vignette Author"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Example Uses of snps2assays}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

There is one main function in this package `assayize()` that takes data frame input
of the necessary ingredients (all variants, variants you want to target, and 
consensus sequences) and then spits out information useful for ordering assays
and knowing whether they have a good chance of being designable.

There are also a few helper functions to get from common formats to the necessary data frames.
For example, from VCF to a data frame of all the variants, or from FASTA to a data frame
of consensus sequences.  We demonstrate those in a series of workflows here.

Let's get going by loading the package:
```{r, message=FALSE}
library(snps2assays)
```

## Some Handy Functions and Example Data Sets

### grab_vcf()
Let's say that you have all your variants in a big gzipped (or not) VCF file.  Here are the
top lines of the example VCF file that gets installed with the `snps2assays` package. They will
run off the page, but you get the idea:
```{r, comment = ""}
# print out the first two lines of data (after all the headers)
con <- gzfile(system.file("textdata", "vcf.txt.gz", package = "snps2assays"))
readLines(con, n = 20) %>%
  cat(sep = "\n")
close(con)
```

To read a VCF file (it can be gzipped, but it can't be a BCF file!) you can use the `grab_vcf` function
which will slurp the whole thing into a `tbl_df` data frame.  
```{r, comment = ""}
grab_vcf(system.file("textdata", "vcf.txt.gz", package = "snps2assays"))
```
Note that this function searches for 
the header by trying to find a line that starts with "#CHROM", so, if your VCF files don't
roll that way, you'll have problems.  

### Target SNPs
To go along with that example VCF file we have 25 example target SNPs:
```{r, warning=FALSE, comment = ""}
example_target_snps
```
They might be a little truncated in the view above.

### grab_fasta()
Let's look at the example fasta file that comes with the package 
and that corresponds to the sequences in 
the example VCF file.
```{r, comment = ""}
# print out the first 6 lines of the fasta file
con <- gzfile(system.file("textdata", "fasta.txt.gz", package = "snps2assays"))
readLines(con, n = 10) %>%
  cat(sep = "\n")
close(con)
```

That is what the fasta file looks like.  Here is how we can read it into a data frame
with `grab_fasta`.
```{r, comment = ""}
grab_fasta(system.file("textdata", "fasta.txt.gz", package = "snps2assays"))
```
The resulting data frame has columns `CHROM` (should correspond to the `CHROM`'s in the VCF file),
and `Seq` (the consensus sequence).


## Putting it All Together

Now, we see how to put those together and toss everything into `assayize()`.
```{r}
# get the vcf file:
vcf <- grab_vcf(system.file("textdata", "vcf.txt.gz", package = "snps2assays"))

# get the fasta file
fasta <- grab_fasta(system.file("textdata", "fasta.txt.gz", package = "snps2assays"))

# get our data frame of target SNPs
data(example_target_snps)

# now, assayize them!

```
